src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Snippets/XmlSnippetParser.ExpansionTemplate.cs

```diff
@@ -0,0 +1,233 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Xml.Linq;
+using Microsoft.AspNetCore.Razor;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.Editor.Razor.Snippets;
+
+internal partial class XmlSnippetParser
+{
+    /// <summary>
+    /// Shamelessly adapted from https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=/src/Editor/VisualStudio/Impl/Snippet/ExpansionTemplate.cs
+    /// with changes to parsing to store the snippet as a set of parts instead of a single string.
+    /// </summary>
+    private class ExpansionTemplate
+    {
+        private record ExpansionField(string ID, string Default, string? FunctionName, string? FunctionParam, bool IsEditable);
+
+        private const string Selected = "selected";
+        private const string End = "end";
+        private const string Shortcut = "shortcut";
+
+        private readonly List<ExpansionField> _tokens = new();
+
+        private readonly string? _code;
+        private readonly string _delimiter = "$";
```

Comment:
Can this ever be more than a single character? If not, could it just be a `char` field. That would simplify a lot of the comparisons below.

Reply:
I believe it can, although in practice I don't know if people do. It's not well documented in https://learn.microsoft.com/en-us/visualstudio/ide/code-snippets-schema-reference?view=vs-2022#attributes

Reply:
Reading the parser below, it looks like only a single char would be supported. I'll update to that